lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -0,0 +1,160 @@
+  using OpRewritePattern<memref::AllocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::AllocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto attr = op->getAttr("no_trans");
+    if (attr && attr.isa<mlir::BoolAttr>() &&
+        attr.cast<mlir::BoolAttr>().getValue()) {
+      success();
+    }
+    // The operation has a "no_trans" attribute set to true
+    auto loc = op->getLoc();
+    MemRefType type = op.getMemref().getType();
+    ValueRange symbolOperands = op.getSymbolOperands();
+    ValueRange dynamicSizes = op.getDynamicSizes();
+    // Check if the parent contains a forall / parallel loop
+    bool hasParallelParent = false;
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newAllocOp;
+    if (hasParallelParent) {
+      newAllocOp = rewriter.create<cpuruntime::ThreadAllocOp>(
+          loc, type, dynamicSizes, symbolOperands);
+    } else {
+      newAllocOp = rewriter.create<cpuruntime::AllocOp>(loc, type, dynamicSizes,
+                                                        symbolOperands);
+    }
+    rewriter.replaceOp(op, newAllocOp->getResults());
+    return success();
+  }
+};
+
+struct AlignedDeallocLowering : public OpRewritePattern<memref::DeallocOp> {
+  using OpRewritePattern<memref::DeallocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::DeallocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    Value memref = op.getMemref();
+    // Check if the parent contains a forall / parallel loop
+    bool hasParallelParent = false;
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newDeallocOp;
+    if (hasParallelParent) {
+      newDeallocOp = rewriter.create<cpuruntime::ThreadDeallocOp>(loc, memref);
+    } else {
+      newDeallocOp = rewriter.create<cpuruntime::DeallocOp>(loc, memref);
+    }
+    rewriter.replaceOp(op, newDeallocOp->getResults());
+    return success();
+  }
+};
+
+struct ConvertMemRefToCPURuntime
+    : public impl::ConvertMemRefToCPURuntimeBase<ConvertMemRefToCPURuntime> {
+
+  void runOnOperation() final {
+    auto *ctx = &getContext();
+    // add lowering target
+    ConversionTarget target(getContext());
+    target.addLegalDialect<
+        // clang-format off
+        BuiltinDialect,
+        func::FuncDialect,
+        memref::MemRefDialect,
+        cpuruntime::CPURuntimeDialect,
+        math::MathDialect,
+        arith::ArithDialect,
+        scf::SCFDialect,
+        tensor::TensorDialect,
+        linalg::LinalgDialect,
```

Comment:
do we really need list all legal dialects?

Reply:
narrow down the legal dialects.